### ISSUE TITLE
Recover from minor performance regressions

### ISSUE THREAD

**Issue by aartbik** [dart-lang/sdk#35323]
The control dep as data dep CL has landed and seems to stick. However, we had a few minor regressions (and one gain!) due to the change, most likely due to missed cases of looking under the hood of a check (but perhaps also due to missed LICM which may be unavoidable). This bug tracks if something can be done.

```
Bench2D | -6.03 % | RunTime | dart-aot | Odroid-C2 | linux-armv7hf | 
Meteor | -4.15 % | RunTime | dart-aot | Intel Xeon | linux-x64 |
ParticleSystemPaint | +8.88 % | RunTime | dart-aot | Intel Core 2 | linux-x64 |
SkeletalAnimation | -5.79 % | RunTime | dart-aot | Intel Core i5 | linux-x64 |
SkeletalAnimation | -7.62 % | RunTime | dart-aot | Intel Xeon | linux-x64 |
SkeletalAnimation | -4.77 % | RunTime | dart-aot | Odroid-C2 | linux-armv8 |
TypedData.Float64ListVarBench | -9.83 % | RunTime | dart-aot | Odroid-C2 | linux-armv7hf |
```

**Comment by aartbik** [comment]
For SkeletalAnimation, the extra data dependence has no adverse impact on performance. The full difference can be explained by lack of LICM on the non-speculative checks (in the past, we would hoist the checks as speculative checks, and then later convert them to non-speculative).

```
AOT:
SkeletalAnimation(RunTime): 74.88025271433919 us.     // allow LICM on non-speculative check
SkeletalAnimation(RunTime): 82.62201896145744 us.     // no LICM on non-speculative (ToT status)
```
Let's see if we can do something even for non-speculative checks (under much stricter conditions of course)....

**Comment by aartbik** [comment]
Latest inliner heuristics works well for most, but not for these:

```
CorelibIsolateStartup | -8.76 % | RunTime | dart | Intel Core i5 | linux-ia32 | -1.04 x larger than benchmark noise
-- | -- | -- | -- | -- | -- | --
CorelibIsolateStartup | -9.99 % | RunTime | dart | Intel Core i5 | linux-x64 | -1.12 x larger than benchmark noise
CorelibIsolateStartup | -8.8 % | RunTime | dart | Intel Xeon | linux-ia32 | -1.18 x larger than benchmark noise
CorelibIsolateStartup | -8.37 % | RunTime | dart | Intel Xeon | linux-x64 | -1.06 x larger than benchmark noise
DartMicroBench.IntPower2 | -10.7 % | RunTime | dart-aot | Odroid-C2 | linux-armv7hf | -3.21 x larger than benchmark noise
DartMicroBench.IntPower3 | -7.76 % | RunTime | dart-aot | Odroid-C2 | linux-armv7hf | -2.01 x larger than benchmark noise
JsonParseDefaultReviver | -10.7 % | RunTime | dart-aot | Intel Core i5 | linux-x64 | -2.27 x larger than benchmark noise
-- | -- | -- | -- | -- | -- | --
JsonParseDefaultReviver | -11.1 % | RunTime | dart-aot | Intel Xeon | linux-x64 | -2.43 x larger than benchmark noise
NextLoop | -4.42 % | RunTime | dart-aot | Intel Core i5 | linux-x64 | -1.27 x larger than benchmark noise
-- | -- | -- | -- | -- | -- | --

```

**Comment by aartbik** [comment]
Also adding Slava's excellent notes on code size increase cases that could be fixed. We may split this out in individual trackers for the ones that need more work, or for other cases we find.

> 
> (1) numberFormatSymbols seems to be a initializer for the static field. Maybe we should tune inlining down in the initializers - because they are executed only once?
> (2) Often the code is polluted by redundant repeated Unbox instructions unboxing the same value. An additional round of CSE should fix this.
> (3) I saw multiple cases where we inline _ByteDataView.[set|get]Xyz methods that look like this
> 
>   int getInt8(int byteOffset) {
>     if (byteOffset < 0 || byteOffset >= length) {
>       throw new RangeError.index(byteOffset, this, "byteOffset");
>     }
>     return _typedData._getInt8(_offset + byteOffset);
>   }
> 
> This adds a lot of code for throwing the RangeError. We should have a way to keep throwing out-of-line. This is probably contributing a lot to the binary size.
> 
> (3.1) I noticed that _ByteDataView fields are not marked with pragma annotations, so we constantly check if they are null, but they are not null by construction. They should be annotated with annotations similar to those on _TypeListView fields:
> 
>   @pragma("vm:non-nullable-result-type")
>   final _TypedList _typedData;
> 
>   @pragma("vm:exact-result-type", "dart:core#_Smi")
>   final int offsetInBytes;
> 
>   @pragma("vm:exact-result-type", "dart:core#_Smi")
>   final int length;
> 
> (3.2) I noticed that unless ByteData construction is inlined we can't forward length. We should add canonicalization rule that does replacement ByteData(n).length => n. Similar we should forward ByteData(n)._offset => 0. 
>  
> (3.3) I noticed that constant folding is confused about boxing and unboxing operations, e.g. UnboxInt64(1) > UnboxInt64(2) does not fold to false as I would expect.

**Comment by aartbik** [comment]
Nothing left here that does not fall under general performance improvements that are ongoing.
